marklogic-data-hub/src/server-side/impl/trace-lib.sjs

```diff
@@ -0,0 +1,375 @@
+/**
+  Copyright 2012-2018 MarkLogic Corporation
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+'use strict';
+
+const config = require("/com.marklogic.hub/config.sjs");
+if (!this.rfc) {
+  this.rfc = require("/MarkLogic/data-hub-framework/impl/run-flow-context.sjs");
+}
+const tracelib = require('/MarkLogic/data-hub-framework/impl/trace-lib.xqy');
+
+let internalContexts = {
+  currentTraceSettings: {}
+}
+
+function setCurrentTraceSettings(settings) {
+  internalContexts.currentTraceSettings = settings;
+}
+
+function getCurrentTraceSettings() {
+  return internalContexts.currentTraceSettings;
+}
+
+function newTrace() {
+  return {
+    traceId: xdmp.random(),
+    created: fn.currentDateTime()
+  };
+}
+
+function enableTracing(enabled) {
+  return tracelib.enableTracing(enabled);
+}
+
+function enabled() {
+  return tracelib.enabled();
+}
+
+function hasErrors() {
+  let ts = getCurrentTraceSettings();
+  return !!ts['_has_errors'];
+}
+
+function incrementErrorCount() {
+  let ts = getCurrentTraceSettings();
+  ts.errorCount = tracelib.getErrorCount() + 1;
+}
+
+function getErrorCount() {
+  let ts = getCurrentTraceSettings();
+  return ts.errorCount || 0;
+}
+
+function addFailedItem(item)
+{
+  getFailedItems().push(item);
+}
+
+function addCompletedItem(item) {
+  getCompletedItems().push(item);
+}
+
+function getCurrentTrace(currentTrace) {
+  let ct = currentTrace || rfc.getTrace(rfc.getItemContext());
+  if (!ct) {
+    fn.error(xs.QName("MISSING_CURRENT_TRACE"));
+  }
+  return ct;
+}
+
+function setPluginLabel(label, currentTrace) {
+  let ct = getCurrentTrace(currentTrace);
+  ct.pluginLabel = label;
+}
+
+function getPluginLabel(currentTrace) {
+  return currentTrace.pluginLabel;
+}
+
+function resetPluginInput(currentTrace) {
+  let ct = getCurrentTrace(currentTrace);
+  ct.pluginInput = {};
+}
+
+function getPluginInput(currentTrace) {
+  let o = currentTrace.pluginInput;
+  if (o) {
+    if (rfc.isJson()) {
+      let oo = {};
+      for (let key in o) {
+        let value = o[key];
+        if (value instanceof BinaryNode) {
+          value = "binary data";
+        }
+        oo[key] = value;
+      }
+      return oo;
+    }
+    else {
+      const x = new NodeBuilder();
+      return x.toNode();
```

Comment:
returning here means the following lines won't be run. Looks like `return x.toNode();` should be after the for loop.

Reply:
looks like an edge case that would break if it were properly tested. I'll fix that.